controller/src/main/java/io/pravega/controller/store/stream/tables/StreamTruncationRecord.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream.tables;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Data class for storing information about stream's truncation point.
+ */
+@AllArgsConstructor(access = AccessLevel.PACKAGE)
+public class StreamTruncationRecord implements Serializable {
+    public static final StreamTruncationRecord EMPTY = new StreamTruncationRecord(ImmutableMap.of(),
+            ImmutableMap.of(), ImmutableSet.of(), ImmutableSet.of());
+
+    /**
+     * Stream cut that is applied as part of this truncation.
+     */
+    private final ImmutableMap<Integer, Long> streamCut;
+
+    /**
+     * If a stream cut spans across multiple epochs then this map captures mapping of segments from the stream cut to
```

Comment:
I'm a bit confused with this discussion here:

1- It is not clear why we need to look at previous epochs. Any segment that was active in previous epoch and is still active will be present in the latest epoch. In the example, segment 0 spans multiple epochs, but ultimately, it is a member of epoch 3. Consequently, all we need to do is to inspect epoch 3.
2- Could you clarify the reference to `TableHelper.getActiveSegmentsAt`? That call actually doesn't exist, so I think you might be referring to `TableHelper.getActiveSegments` that takes a timestamp as an argument.

Reply:
1. take for example the following scenario:
time: epoch : segments 
T0    : 0        : 0 1
T1    : 1        : 2, 3, 4, 5  // 0 -> 2, 3; 1 -> 4, 5
T3    : 2        : 2, 3, 6     // 4, 5 -> 6

Now readers in a reader group could be reading from segment 0 and 6. And that is a perfectly valid read case. So when they create a cut, this streamCut will span across 6 epochs namely 0 to 2. 

So when we compute segments to delete for this stream cut, we can delete all predecessors of 6 and 0. Which means we can delete 4, 5 and 1. 
To identify 4, 5, and 1 we need to traverse back one epoch at a time and identify how the stream cut impacts segments in the epoch and compute predecessors. 

Also note, get active segments at different times become interesting
Between  0  - T0: 0, 6
Between T0 - T1: 0, 6 
Between T1 - T2: 2, 3, 6
After T2:            : 2, 3, 6

2. yes

Reply:
Ok.